File: stack/155/155.py
# solution 1
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []

    def push(self, x: int) -> None:
        currMin = self.getMin()
        if currMin is None or x < currMin:
            currMin = x
        self.stack.append((x, currMin))

    def pop(self) -> None:
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1][0]

    def getMin(self) -> int:
        if len(self.stack) == 0:
            return None
        else:
            return self.stack[-1][1]


# solution 2
class Node:
    def __init__(self, x, min, next):
        self.val = x
        self.currMin = min
        self.next = next


class MinStack:
    def __init__(self):
        self.head = None

    def push(self, x):
        currMin = self.getMin()
        if currMin is None or x < currMin:
            currMin = x
        self.head = Node(x, currMin, self.head)

    def pop(self):
        topVal = self.head.val
        self.head = self.head.next
        return topVal

    def top(self):
        return self.head.val

    def getMin(self):
        return None if self.head is None else self.head.currMin

File: stack/155/155.js
// solution 1
class MinStack {
  constructor() {
    this.stack = [];
  }

  push(x) {
    let currMin = this.getMin();
    if (currMin === null || x < currMin) {
      currMin = x;
    }
    this.stack.push([x, currMin]);
  }

  pop() {
    this.stack.pop();
  }

  top() {
    return this.stack[this.stack.length - 1][0];
  }

  getMin() {
    if (this.stack.length === 0) {
      return null;
    } else {
      return this.stack[this.stack.length - 1][1];
    }
  }
}

// solution 2
class Node {
  constructor(x, currMin, next) {
    this.val = x;
    this.currMin = currMin;
    this.next = next;
  }
}
class MinStack {
  constructor() {
    this.head = null;
  }
  push(x) {
    let currMin = this.getMin();
    if (currMin === null || x < currMin) {
      currMin = x;
    }
    this.head = new Node(x, currMin, this.head);
  }
  pop() {
    const topVal = this.head.val;
    this.head = this.head.next;
    return topVal;
  }
  top() {
    return this.head.val;
  }
  getMin() {
    return this.head ? this.head.currMin : null;
  }
}

//solution 3
class MinStack {
  constructor() {
    this.stack = [];
    this.mins = [Infinity];
  }

  push(x) {
    if (x <= this.getMin()) {
      this.mins.push(x);
    }
    this.stack.push(x);
  }
  pop() {
    const elem = this.stack.pop();
    if (elem === this.getMin()) {
      this.mins.pop();
    }
  }
  top() {
    return this.stack[this.stack.length - 1];
  }

  getMin() {
    return this.mins[this.mins.length - 1];
  }
}
